Speech commands use sequences

feerrenrut opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
#10098

### Summary of the issue:
In order to get many of the benefits of the speech refactor, we need to have speech sequences (rather than single strings) returned from our various `getXSpeech` methods. This will be necessary if we want to use sounds instead of text to indicate spelling errors, change voice parameters for emphasized text or links, etc.

### Description of how this pull request fixes the issue:
The following functions have had their definitions and usages converted:
- `getControlFieldSpeech`
- `speech.getControlFieldSpeech`
  - `browseMode.BrowseModeDocumentTextInfo.getControlFieldSpeech`
  - `textInfos.TextInfo.getControlFieldSpeech`
  - `speech.getFormatFieldSpeech`
- `getFormatFieldSpeech`
  - `appModules.kindle.BookPageViewTextInfo.getFormatFieldSpeech`
  - `textInfos.TextInfo.getFormatFieldSpeech`
  - `treeInterceptorHandler.RootProxyTextInfo.getFormatFieldSpeech`
- `speech.getSpeechTextForProperties`
- `speech.getIndentationSpeech`
- `speech.getTableInfoSpeech`

The speech function `getSpeechTextForProperties` has been renamed to `getPropertiesSpeech`

### Testing performed:
Generally ran NVDA and inspected the log for errors:
- In Chrome and Firefox
- Excel, Word
- Start menu

### Known issues with pull request:

### Change log entry:

Section: Changes For Developers
```
Several speech functions have been changed to return speech sequences
- getControlFieldSpeech
- getFormatFieldSpeech
- getSpeechTextForProperties / getPropertiesSpeech
- getIndentationSpeech
- getTableInfoSpeech
```
`The speech function getSpeechTextForProperties has been renamed to getPropertiesSpeech`

LeonarddeR committed:
Speech functions now return a SpeechSequence

Definitions and all usages converted:

- getControlFieldSpeech
  - speech.getControlFieldSpeech
  - browseMode.BrowseModeDocumentTextInfo.getControlFieldSpeech
  - textInfos.TextInfo.getControlFieldSpeech
  - speech.getFormatFieldSpeech
- getFormatFieldSpeech
  - appModules.kindle.BookPageViewTextInfo.getFormatFieldSpeech
  - textInfos.TextInfo.getFormatFieldSpeech
  - treeInterceptorHandler.RootProxyTextInfo.getFormatFieldSpeech
- speech.getSpeechTextForProperties
- speech.getIndentationSpeech
- speech.getTableInfoSpeech

feerrenrut committed:
Rename getSpeechTextForProperties to getPropertiesSpeech

feerrenrut committed:
Unify line endings (use Windows line endings)

feerrenrut commented on source/appModules/kindle.py:
```diff
@@ -272,29 +271,42 @@ def getTextWithFields(self, formatConfig=None):
-		out = ""
+	def getFormatFieldSpeech(
+			self, attrs, attrsCache=None, formatConfig=None, reason=None, unit=None, extraDetail=False,
+			initialFormat=False, separator=speech.CHUNK_SEPARATOR
```
Param `separator` can be removed. Anyone disagree with this?

feerrenrut commented on source/textInfos/__init__.py:
```diff
@@ -472,17 +477,26 @@ def getTextInChunks(self, unit):
-	def getFormatFieldSpeech(self, attrs, attrsCache=None, formatConfig=None, reason=None, unit=None, extraDetail=False , initialFormat=False, separator=None):
+	def getFormatFieldSpeech(
+			self, attrs, attrsCache=None, formatConfig=None, reason=None, unit=None, extraDetail=False,
+			initialFormat=False, separator=None
```
Param separator can be removed. Anyone disagree with this?

leonardder replied:
I"m not entirely sure. I think this was introduced because some speech synthesizers require double spaces between strings to make sure numbers are announced correctly.

feerrenrut commented on source/speech/__init__.py:
```diff
@@ -1019,13 +1076,22 @@ def speakTextInfo(info, useCache=True, formatConfig=None, unit=None, reason=cont
 	if autoLanguageSwitching and lastLanguage is not None:
-		speechSequence.append(LangChangeCommand(None))
+		speechSequence.append(
+			LangChangeCommand(None)
```
I don't think this should be None!

feerrenrut replied:
I wasn't so sure, the commented out line above made me question this: `self.lang=lang # if lang else languageHandler.getLanguage()` . Looking into it further 
`command.isDefault` is only used in `ParamChangeTracker.update`, and here it removes any existing `LangChangeCommand` from `_commands`. What isn't clear to me is if there is state stored elsewhere. Does a subsequent command need to be issued to return the language to the system default?

I think generally this isn't something to change in this PR, but it made me wonder.

leonardder commented on source/NVDAObjects/window/excel.py:
```diff
@@ -229,13 +229,13 @@ class ExcelChartQuickNavItem(ExcelQuickNavItem):
+		if chartObject.Chart.HasTitle:
 
-			self.label = chartObject.Chart.ChartTitle.Text + " " + chartObject.TopLeftCell.address(False,False,1,False) + "-" + chartObject.BottomRightCell.address(False,False,1,False) 
+			self.label = chartObject.Chart.ChartTitle.Text + " " + chartObject.TopLeftCell.address(False,False,1,False) + "-" + chartObject.BottomRightCell.address(False,False,1,False) 
```
Do you know why this ended up in the diff?

feerrenrut replied:
yes, there is a commit that converts lines in that file to windows lines. It is the last commit, you can exclude it from the diff.

feerrenrut commented:
One other thing I noticed will doing this. Refer to line 239 in speech.manager
https://github.com/nvaccess/nvda/blob/cfbf173ec99b87ebbfb00ca20e01afbedcba3ad6/source/speech/manager.py#L239

All other `isinstance` checks end with continue, why is this one different?
Should there be a default "catch-all" here? Unknown commands? Unaccepted types?

nvdaes commented:
Not sure if issue #10363 is related to this. In this case, as an alternative, characterProcessing.processSpeechSymbols may return a sequence too?

leonardder commented on source/speech/__init__.py:
```diff
@@ -1244,31 +1328,51 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
-			stateText, 
-			getSpeechTextForProperties(
+		tableSeq = nameSequence
+		tableSeq.extend(roleTextSequence)
```
Be aware of this gotcha:

```
>>> f=[]
>>> f.extend("hello")
>>> print(f)
['h', 'e', 'l', 'l', 'o']
```

This is exactly what's happening now with roleText if it is specifically set, e.g. aria-roledescription.

`data:text/html, <p>This is a <button aria-roledescription="fancy button">fancy button label</button>`

The fancy button role is announced as `f a n c y   b u t t o n`

leonardder commented on source/speech/__init__.py:
```diff
@@ -1244,31 +1328,51 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
-			roleText,
-			stateText, 
-			getSpeechTextForProperties(
+		tableSeq = nameSequence
```
Note that this also touches the nameSequence when extending. Also for uniformity, I'd change its name.

```suggestion
		tableSequence = nameSequence[:]
```

leonardder commented on source/speech/__init__.py:
```diff
@@ -1282,51 +1386,89 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
-		return (getSpeechTextForProperties(_tableID=tableID, **getProps)
-			+ (" %s" % stateText if stateText else "")
-			+ (" %s" % ariaCurrentText if ariaCurrent else ""))
+		tableCellsequence = getPropertiesSpeech(_tableID=tableID, **getProps)
```
Typo

```suggestion
		tableCellSequence = getPropertiesSpeech(_tableID=tableID, **getProps)
```

leonardder commented on source/speech/__init__.py:
```diff
@@ -1282,51 +1386,89 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
+		return [
 		# Translators: Indicates end of something (example output: at the end of a list, speaks out of list).
-		return _("out of %s")%roleText
+			_("out of %s") % " ".join(roleTextSequence)
```
This will break if roleTextSequence contains speech commands.

feerrenrut replied:
I'm not really sure what to do here. We could prepend the string "out of " and extend with roleTextSequence, but the "out of " string really is meaningless without subsequent text. Is there a way to do this, I'm perhaps jumping to conclusions, but I wonder if speech sequences should actually be "tree" like.  I'm probably missing something, @jcsteh would you be able to clarify for me?

leonardder replied:
With tree like, do you mean having a list of strings/commands in another list that could either contain more lists or strings or commands?

feerrenrut replied:
> With tree like, do you mean having a list of strings/commands in another list that could either contain more lists or strings or commands?

Yes, essentially. So that you can bundle groups of related speech text or commands (due to them having the same origin / context).  Perhaps we could do this with commands, perhaps this is the intention behind the `endUtteranceCommand`

jcsteh replied:
The "out of" indicator just doesn't work for roles which aren't described textually. For example, if we use a click sound for a link, "out of <click>" is just ridiculous. If the role doesn't produce text, we should just skip this altogether.

One question is what to do if the role has both text and speech commands. One could argue we should take both, but I don't think this is a good idea. For now, I think we should just only do this "out of" thing for roles that produce only text. In future, we may be able to come up with other options; e.g. if a role produces sound, we could have an "out of" sound so we end up with "<out of sound> <role sound>". I think that's getting way ahead of ourselves here though.

leonardder commented on source/speech/__init__.py:
```diff
@@ -1736,14 +1878,19 @@ def getFormatFieldSpeech(attrs,attrsCache=None,formatConfig=None,reason=None,uni
 
-def getTableInfoSpeech(tableInfo,oldTableInfo,extraDetail=False):
+def getTableInfoSpeech(
+		tableInfo, oldTableInfo, extraDetail=False
```
Could you please type annotate these while at it? It's not directly clear to me what type is expected here.

leonardder commented on source/textInfos/__init__.py:
```diff
@@ -13,11 +13,16 @@
 import locationHelper
 
+
+SpeechSequence = List[Union[Any, str]]
```
Probably al
```suggestion
from speech.types import SpeechSequence
```

feerrenrut replied:
I think doing this results in a circular import

leonardder commented on source/textInfos/__init__.py:
```diff
@@ -472,17 +477,26 @@ def getTextInChunks(self, unit):
 
-	def getControlFieldSpeech(self, attrs, ancestorAttrs, fieldType, formatConfig=None, extraDetail=False, reason=None):
+	def getControlFieldSpeech(
+			self, attrs, ancestorAttrs, fieldType, formatConfig=None, extraDetail=False, reason=None
```
If AttrsType is in speech.types, this can be annotated as well.

leonardder commented on source/textInfos/__init__.py:
```diff
@@ -472,17 +477,26 @@ def getTextInChunks(self, unit):
+		sequence = speech.getControlFieldSpeech(
+			attrs, ancestorAttrs, fieldType, formatConfig, extraDetail, reason
+		)
+		# remove empty strings or empty sequences which originate from speech.getControlFieldSpeech
```
I guess we need to do this here because `# remove empty strings or empty sequences which originate from speech.getControlFieldSpeech` has some early returns?

feerrenrut replied:
I'm not entirely sure. I might have to go back and look at this again. It may have been a result of only being partially finished converting functions.

LeonarddeR reviewed (changes requested):
Currently this goes pretty wild on Github with firefox, it announces sections, groupings, property pages in situations I don't expect it to. Could find out that quickly where the bug is, probably somewhere in getControlFieldSpeech.

feerrenrut commented on source/aria.py:
```diff
@@ -3,10 +3,11 @@
 #Copyright (C) 2009-2014 NV Access Limited
```
I just added some type hints, these eventually end up impacting the type hints in the various getXSpeech methods.

feerrenrut commented on source/browseMode.py:
```diff
@@ -1104,23 +1108,41 @@ def move():
 			core.callLater(100, move)
 
+
+def _checkType(value):
```
Errors were occurring via here about sequences containing lists and empty strings. Possibly safe to remove now, or maybe it should be everywhere? I could put these checks behind a debugging flag.

feerrenrut commented on source/appModules/kindle.py:
```diff
@@ -272,29 +271,42 @@ def getTextWithFields(self, formatConfig=None):
-	def getFormatFieldSpeech(self, attrs, attrsCache=None, formatConfig=None, reason=None, unit=None, extraDetail=False , initialFormat=False, separator=speech.CHUNK_SEPARATOR):
-		out = ""
+	def getFormatFieldSpeech(
+			self, attrs, attrsCache=None, formatConfig=None, reason=None, unit=None, extraDetail=False,
```
Not necessarily more difficult to read either way (though I'm sure plenty of people have strong preferences). I was only trying to minimise the diff and keep the linter happy.

feerrenrut commented on source/speech/__init__.py:
```diff
@@ -1200,18 +1277,25 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
-	descriptionText=(getSpeechTextForProperties(reason=reason,description=description)
-		if config.conf["presentation"]["reportObjectDescriptions"] else "")
-	levelText=getSpeechTextForProperties(reason=reason,positionInfo_level=level)
+	roleTextSequence = attrs.get('roleText', None)
```
Ah, yes. This was a mistake!

feerrenrut committed:
Passing None to LangChangeCommand is accepted

feerrenrut committed:
Add type hints to help track down issues with sequence building

feerrenrut committed:
Add speechSequence checking

Log errors in speech sequences (unknown types, None entries, empty list entries)

Must be enabled in the advanced panel, logging options.

feerrenrut committed:
Fix several found issues with building speech sequences.

feerrenrut committed:
Fix lint errors

feerrenrut committed:
Other review actions.

feerrenrut commented:
Looks like system tests are failing. From the output it seems to be to do with the separator that we are no longer using.

feerrenrut commented:
Seems like the only cases that doesn't use `CHUNK_SEPARATOR = "  "` is in `_reportFormattingHelper` (`source/globalCommands.py`) where `separator="\n"` is used instead.

The definition of `CHUNK_SEPARATOR` the following comment:
```
#: The string used to separate distinct chunks of text when multiple chunks should be spoken without pauses.
# #555: Use two spaces so that numbers from adjacent chunks aren't treated as a single number
# for languages such as French and German which use space as a thousands separator.
```

The `CHUNK_SEPARATOR` is appended to each string (when not in character mode) at the end of the `speak` method in `source/speech/__init__.py`

This was introduced in 7f870737ab, which introduces browseable messages. It may be necessary to separate text with new lines for the browseable message viewer?

feerrenrut committed:
Remove separator argument from getFormatFieldSpeech method

feerrenrut committed:
Include padding added to strings for speech in test

In the future it would be better to test speech as structured data. This
would allow us to test commands as well.

feerrenrut commented:
This is ready for testing: [try build](https://ci.appveyor.com/api/buildjobs/v0cnl2ceau1dwlnx/artifacts/output%2Fnvda_snapshot_try-i10098-18938%2Cde1343e1.exe)

I wouldn't suggest starting to review this again until we get some confirmation from testers. The commits are a bit of a jumble at this stage. I think there is value in making these changes easy to track, so I may rebase to clean up (re-order) the commits, then do a regular merge rather than squash merge.

nvdaes commented:
I'm testing the try build and I find that, in browse mode, NVDA reports "paragraph" and "section" too frequently. For example, reading the last email corresponding to this PR on in Google's home page.
Congrats for this wonderful work for speech.

leonardder commented on source/speech/__init__.py:
```diff
@@ -887,21 +959,36 @@ def speakTextInfo(info, useCache=True, formatConfig=None, unit=None, reason=cont
 				presCat=field.getPresentationCategory(newControlFieldStack[0:count],formatConfig,reason)
 				if not presCat or presCat is field.PRESCAT_LAYOUT:
```
Just a suggestion:

Initially, I thought that the presentation category code might cause the section reporting, but it is untouched. Yet, I'd like to point your attention at it. `textInfos.ControlField.PRESCAT_LAYOUT` is None, which to me is pretty confusing. may be it should be changed to `"layout"` so we can use proper type checking to avoid issues now or in the future.

feerrenrut replied:
This is a good point. I would be with you on this idea, I'd be interested to know what @jcsteh thinks though? 

Jamie:
In short the value for `textInfos.ControlField.PRESCAT_LAYOUT` is `None`, how risky would it be to give this an actual default value like "layout"? Is there a reliance on not being able to retrieve an attribute or dictionary value and therefore `None` matches the intended default?

Happy for the answer to be "Not sure, please investigate" but if you know off-hand it might save us time.

LeonarddeR reviewed (commented):
Not really reviewing, just trying to find out what causes the reporting of roles.

feerrenrut commented:
@nvdaes Thanks for testing

> in browse mode, NVDA reports "paragraph" and "section" too frequently.

Could you confirm whether you see the same behaviour with the latest alpha?

@leonardder 
> this goes pretty wild on Github with firefox, it announces sections, groupings, property pages in situations I don't expect it to

I went looking for strange behaviour on Github with firefox with the most recent commit. I found some things I didn't expect, but this is also the case using 2019.2.

feerrenrut commented:
> textInfos.ControlField.PRESCAT_LAYOUT is None, which to me is pretty confusing. may be it should be changed to "layout" so we can use proper type checking to avoid issues now or in the future.

I agree in theory, though by having a value of None the result might be that it is acting as a default when nothing else is set. We would need to track down all these cases and make it explicit.

leonardder commented:
On alpha, when I do ctrl+end from the comment edit box in browse mode to get to the bottom of this github page, NVDA says: ``content info landmark visited  link About`

With the code of this pr, the announcement is: `content info landmark unknown list item visited link About`

The `unknown` and `list item` roles shouldn't be announced.

The speech viewer also shows several unexpected empty lines.

feerrenrut commented:
> With the code of this pr, the announcement is: `content info landmark unknown list item visited link About`

I am having trouble reproducing this (using Firefox):
I move to comment field, NVDA announces:
```
Comment body
edit
required
multi line
blank
```

I press control+end, NVDA announces:
```
out of edit
section

shows several unexpected empty lines.
```

The most interesting part to me is:

> shows several unexpected empty lines.

This is from the end of your comment, I have no idea how that is ending up in the announcement.

feerrenrut commented:
> I am having trouble reproducing this (using Firefox):

My mistake, I have too many windows open, that result was from Chrome, I will edit the comment.

In Firefox 69.0.3 (64-bit), I get the following after pressing ctrl+end:
```
out of edit
content info landmark
section

list
with 6 items
list item

link

About
```

nvdaes commented:
Just to answer your question:
feerrenrut commented 5 hours ago 
@nvdaes Thanks for testing
in browse mode, NVDA reports "paragraph" and "section" too frequently.
Could you confirm whether you see the same behaviour with the latest alpha?

No, this is specific of this PR.

leonardder commented:
Ugh, I'm sorry. I intended to say that I see several empty lines in the 
speech viewer within speech sequences where I don't see them on alpha.


Anyway, the fact that your system announces 'section' is already wrong, 
section should never be announced in browse mode.

leonardder commented:
This should be an appropriate test case:

```
data:text/html, <div>First section</div><p>First paragraph</p><p>Second paragraph</p><div>Second section</div>
```

Here, the roles are announced in Firefox browse mode with this pr. However... they shouldn't be.

feerrenrut commented:
@leonardder Thanks for the test case, I'll investigate this shortly.

feerrenrut committed:
Fix speak role too often in browse mode

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into i10098-speechCommandsUseSequences

Resolved Conflicts:
	source/aria.py
	source/browseMode.py
	source/speech/__init__.py

feerrenrut committed:
Fix up lint errors.

feerrenrut commented:
I believe I have addressed the issues described. Please [try out the build](https://ci.appveyor.com/api/buildjobs/x9kq92cbg1gr5j0e/artifacts/output%2Fnvda_snapshot_try-i10098-19075%2C9bbc10b8.exe)

leonardder commented on source/speech/manager.py:
```diff
@@ -151,6 +151,8 @@ def _reset(self):
 
 	def speak(self, speechSequence, priority):
+		if "section" in speechSequence and "First section" in speechSequence:
+			log.error("Found it")
```
Looks like this is commited accidentally.

feerrenrut replied:
Haha, yes I'd say so. Interestingly, this isn't how I found the error.

feerrenrut committed:
Fix: remove debug logging

leonardder commented:
Working on #10462, I realized that braille contains the getBrailleTextForProperties function. Could we consider renaming this to getPropertiesBraille, in line with the rename in the speech package? There doesn't have to change anything else for braille, but I think that renaming one without the other could lead to confusion or at least introduces an inconsistency.

feerrenrut commented:
> Could we consider renaming this to getPropertiesBraille, in line with the rename in the speech package?

I agree, this would be a good idea. Next question is where to do it? I think my preference is to follow this PR with another one specifically to rename it.

leonardder commented:
> I think my preference is to follow this PR with another one 
specifically to rename it.AGreed.

LeonarddeR reviewed (changes requested):
The behavior of the speech viewer still differs:

Old situation:

```
Show Speech Viewer on Startup  check box  not checked  Alt+s
```

New situation:

```
Show Speech Viewer on Startup
check box
not checked
Alt+s
```

Turns out that every chunk of speech is written to a new line.

leonardder commented:
Apart from the speech viewer issue mentioned above, I've been using the branch from source for some time now without issues.

feerrenrut commented:
> Turns out that every chunk of speech is written to a new line.

Initially, my thought was this would be helpful, and it was for debugging purposes. From a usability standpoint, it's probably overall better if sequences would be joined by a space. Though there are certainly cases where people get confused looking at the speechviewer thinking that information is double speaking. I think I would prefer to solve that problem with more formatting in the speechviewer.

> I've been using the branch from source for some time now without issues.

Great!

feerrenrut commented on source/config/configSpec.py:
```diff
@@ -221,6 +221,7 @@
 	louis = boolean(default=false)
 	timeSinceInput = boolean(default=false)
 	vision = boolean(default=false)
+	speechSequences = boolean(default=false)
```
Seems sensible to me.

feerrenrut committed:
Rename "speechSequences" logging category to "speech"

leonardder commented:
> Initially, my thought was this would be helpful, and it was for debugging purposes.

Of course, I don't know how it looks visually. I only use the speech viewer to be able to copy speech sequences (also for debugging purposes), and in that case it's a bit confusing. I'd also separate every sequence with an empty line to make it visible where one sequence ends and the other starts, in that case.

feerrenrut committed:
Fix unnecessary newlines between speech items in Speech viewer

SpeechViewer.appendText renamed to appendSpeechSequence.
It takes a SpeechSequence instead of a str.

This allows more control over the presentation of speech, which will
separate speech items with a space, and conclude with a newline.

feerrenrut commented:
> I'd also separate every sequence with an empty line

I think this will make speech viewer too sparse. Especially considering how often newlines are already in the speech sequence. What I have now should restore the old behaviour, I think these other ideas should be discussed separately.

leonardder commented on source/mathPres/__init__.py:
```diff
@@ -31,7 +33,7 @@ def getSpeechForMathMl(self, mathMl):
 		@type mathMl: str
 		@return: A speech sequence.
-		@rtype: list of str and/or L{speech.SpeechCommand}
+		@rtype: List[str, speech.SpeechCommand]
```
This can now be type annotated I assume

feerrenrut replied:
No, I don't think so.  Because speech imports mathPres. Hopefully we eventually move things around so we stop running into these issues.

leonardder commented on source/speech/__init__.py:
```diff
@@ -7,29 +7,52 @@
+	EndUtteranceCommand,
+	CharacterModeCommand,
+)
+from .commands import (  # noqa: F401
```
Nice improvement!

leonardder commented on source/speech/__init__.py:
```diff
@@ -7,29 +7,52 @@
+	WaveFileCommand,
+	ConfigProfileTriggerCommand,
+)
+import types
```
I realized that this is now a bit ambiguous, as we have types (a built-in python module) and speech.types. I assume you're referring to the built-in one here?

```suggestion
import types  # Python core module
```

leonardder commented on source/speech/__init__.py:
```diff
@@ -462,22 +483,27 @@ def splitTextIndentation(text):
-def getIndentationSpeech(indentation, formatConfig):
+
+
+def getIndentationSpeech(indentation, formatConfig) -> SpeechSequence:
```
```suggestion
def getIndentationSpeech(indentation: str, formatConfig: Dict[str, bool]) -> SpeechSequence:
```

leonardder commented on source/speech/__init__.py:
```diff
@@ -506,26 +532,33 @@ def getIndentationSpeech(indentation, formatConfig):
+def speak(  # noqa: C901
+		speechSequence: SpeechSequence,
+		symbolLevel: Optional[int] = None,
+		priority: Optional[int] = None
```
I'm not sure whether I suggested this before, but as speech priorities are new to speech refactor and thus new to the Python 3 version of NVDA, we could consider making this a enum.IntEnum. Not as part of this pr, though, but I'm happy to provide a pr accordingly.

feerrenrut replied:
That sounds like a good idea if you have time for it 🥇

leonardder commented on source/speech/__init__.py:
```diff
@@ -790,7 +826,25 @@ def _speakTextInfo_addMath(speechSequence, info, field):
-def speakTextInfo(info, useCache=True, formatConfig=None, unit=None, reason=controlTypes.REASON_QUERY, _prefixSpeechCommand=None, onlyInitialFields=False, suppressBlanks=False,priority=None):
+
+def _logBadSequenceTypes(sequence, shouldRaise: bool = True) -> bool:
+	return types.logBadSequenceTypes(sequence, raiseExceptionOnError=shouldRaise)
```
Ah, so it turns out that at the top of the file, my assumption that you were importing Python core types module was incorrect. That has to be from . import types, then.

leonardder commented on source/speech/__init__.py:
```diff
@@ -790,7 +826,25 @@ def _speakTextInfo_addMath(speechSequence, info, field):
 
-def speakTextInfo(info, useCache=True, formatConfig=None, unit=None, reason=controlTypes.REASON_QUERY, _prefixSpeechCommand=None, onlyInitialFields=False, suppressBlanks=False,priority=None):
+
+def _logBadSequenceTypes(sequence, shouldRaise: bool = True) -> bool:
```
Why is this wrapper here? It seems a bit overkill, why not use speech.types.logBadSequenceTypes directly?

feerrenrut replied:
It just made it easier to change how logging was handled in all of speech. Add extra debugging etc. I'll remove it now.

leonardder commented on source/speech/__init__.py:
```diff
@@ -874,29 +928,45 @@ def speakTextInfo(info, useCache=True, formatConfig=None, unit=None, reason=cont
+			fieldSequence = info.getControlFieldSpeech(
+				controlFieldStackCache[count],
+				controlFieldStackCache[0:count],
+				"end_removedFromControlFieldStack",
```
Ugh, these have to be part of an enum at some point too.

leonardder commented on source/speech/__init__.py:
```diff
@@ -1003,9 +1058,11 @@ def speakTextInfo(info, useCache=True, formatConfig=None, unit=None, reason=cont
 			# but the initial format field specified a different language.
 			# Insert the indentation before the LangChangeCommand.
-			speechSequence.insert(-1, indentationSpeech)
+			langChange = speechSequence.pop()
```
I actually realized that this can be done in one line instead of three.

```suggestion
			speechSequence[-1:-1] = = indentationSpeech)
```

Example:

```
>>> l=[1, 2, 3]
>>> l[-1:-1] = [4,5,6]
>>> l
[1, 2, 4, 5, 6, 3]
```

leonardder commented on source/speech/__init__.py:
```diff
@@ -1071,10 +1186,17 @@ def speakTextInfo(info, useCache=True, formatConfig=None, unit=None, reason=cont
 	global oldTreeLevel, oldTableID, oldRowNumber, oldRowSpan, oldColumnNumber, oldColumnSpan
-	textList=[]
-	name=propertyValues.get('name')
+	textList: List[str] = []
```
This is now a list of strings, but strictly spoken, it may contain commands in the future. If we introduce commands here, we would need to rename textList to sequence or something. Probably best to do this now?

feerrenrut replied:
This is true, but it's introducing unnecessary changes and hopefully having to change the name when introducing commands will encourage more thought into the change in type that is occuring.

leonardder commented on source/speech/__init__.py:
```diff
@@ -1321,54 +1536,110 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
+) -> SpeechSequence:
 	if not formatConfig:
 		formatConfig=config.conf["documentFormatting"]
 	textList=[]
```
Same as above

leonardder commented on source/textInfos/__init__.py:
```diff
@@ -501,32 +512,54 @@ def getTextInChunks(self, unit):
-			# but that can't be specified as a default argument because of circular import issues.
-			separator = speech.CHUNK_SEPARATOR
-		return speech.getFormatFieldSpeech(attrs, attrsCache=attrsCache, formatConfig=formatConfig, reason=reason, unit=unit, extraDetail=extraDetail , initialFormat=initialFormat, separator=separator)
+		return speech.getFormatFieldSpeech(
```
May be you want to log bad sequences here as well, as in getControlFieldSpeech?

feerrenrut replied:
speech.getFormatFieldSpeech logs it. If this method augmented the sequence it should probably check. For now I think this is fine.

leonardder commented on tests/system/NVDA Core/variables.py:
```diff
@@ -3,11 +3,11 @@
-	"Exit NVDA  dialog\n"
+	"Exit NVDA  dialog  \n"
 	"What would you like to do?  combo box  Exit  collapsed  Alt plus d"
 )
```
No line at end of file warning

LeonarddeR reviewed (commented):
I had another look through this. I think it looks pretty good, just some small comments and thoughts.

feerrenrut committed:
Fix wrong "types" import.

feerrenrut committed:
Fix type annotations for getIndentationSpeech

feerrenrut committed:
Remove unnecessary logBadSequenceTypes wrapper

feerrenrut committed:
Fix missing EOF newline

LeonarddeR reviewed (approved):
Github insisted on hiding https://github.com/nvaccess/nvda/pull/10371/files#r342995158 . Doesn't hold me back from approving this, though.

Unless @michaelDCurran wants to review this, I'd like to suggest merging this ASAP for broader testing.
